pkg/resource/clusteroperator.go

```diff
@@ -217,33 +218,44 @@ func isDeploymentStatusAvailableAndUpdated(deploy *appsapi.Deployment) bool {
 		deploy.Status.UpdatedReplicas == deploy.Status.Replicas
 }
 
-func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (modified bool, err error) {
-	deploy, err := gco.deployLister.Get(imageregistryv1.ImageRegistryName)
-	if err != nil {
-		if kerrors.IsNotFound(err) {
-			err = nil
-		}
-		return
-	}
+// syncVersions updates versions reported through ClusterOperator resource.
+//
+// We report two distinct versions, one for the Operator and the other for the
+// Image Registry, we read the first from the environment and the latter from
+// an annotation on Registry's deployment.
+func (gco *generatorClusterOperator) syncVersions(op *configapi.ClusterOperator) (bool, error) {
+	var newVersions []configapi.OperandVersion
 
-	deploymentVersion := deploy.Annotations[imageregistryv1.VersionAnnotation]
-	if len(deploymentVersion) == 0 || !isDeploymentStatusAvailableAndUpdated(deploy) {
-		return
+	if v := os.Getenv("RELEASE_VERSION"); len(v) > 0 {
+		newVersions = append(newVersions, configapi.OperandVersion{
+			Name:    "operator",
+			Version: v,
+		})
 	}
 
-	newVersions := []configapi.OperandVersion{
-		{
-			Name:    "operator",
-			Version: deploymentVersion,
-		},
+	deploy, err := gco.deployLister.Get(imageregistryv1.ImageRegistryName)
+	if err != nil && !kerrors.IsNotFound(err) {
+		// Return error if other than "NotFound".
+		return false, err
+	} else if err == nil {
+		v := deploy.Annotations[imageregistryv1.VersionAnnotation]
+		if len(v) == 0 || !isDeploymentStatusAvailableAndUpdated(deploy) {
+			return false, nil
+		}
+
+		newVersions = append(newVersions, configapi.OperandVersion{
+			Name:    "image-registry",
+			Version: v,
+		})
```

Comment:
I think on upgrade we can have a state where we report `operator` version, but nothing for the `image-registry` even though we have an existing registry there.

I propose the following for this sync function:

1. Convert the current `op.Status.Versions` slice into a `map[string]string` (map[name]version)
2. Always update the `operator` key with the current version
3. Update the `image-registry` key if the deployment has the right annotation and has rolled out.
4. Convert the map back to a `Versions` slice.

Reply:
I got the idea. The approach of converting from `map[string]string` back into a `[]OperandVersion` is incompatible with `reflect.DeepEqual` though, the resulting slice will certainly contain all Versions but there is no guarantee in what order they are going to show up on the new slice: https://blog.golang.org/go-maps-in-action#TOC_7.

I implemented the same behavior in a different way, ptal.